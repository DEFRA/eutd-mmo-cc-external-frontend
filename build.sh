#/bin/bash
curl -O https://nodejs.org/dist/v12.18.4/node-v12.18.4-linux-x64.tar.xz
tar -xf node-v12.18.4-linux-x64.tar.xz --directory /tmp
export PATH="/tmp/node-v12.18.4-linux-x64/bin:$PATH"
rm node-v12.18.4-linux-x64.tar.xz
npm ci
npm run test-vstack
rm -rf /tmp/node-v12.18.4-linux-x64