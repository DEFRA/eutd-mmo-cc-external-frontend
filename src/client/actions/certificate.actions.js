import { isError, showFullPageError } from './index';

export const GET_SUMMARY_DOCUMENT = 'GET_SUMMARY_DOCUMENT';
export const SHOW_SUMMARY_DOCUMENT_ERRORS = 'SHOW_SUMMARY_DOCUMENT_ERRORS';
export const GET_SUMMARY_UNAUTHORISED = 'GET_SUMMARY_DOCUMENT_UNAUTHORISED';

export function showSummaryCertificateErrors(data) {
  return { type: SHOW_SUMMARY_DOCUMENT_ERRORS, payload: { validationErrors: data } };
}

export const getSummaryCertificate = (journey, documentNumber = undefined) => async (dispatch, getState, { orchestrationApi }) => {
  const config = (documentNumber)
    ? { headers: { documentnumber: documentNumber } }
    : {};

  await orchestrationApi.get(`/certificate/${journey}`, config)
    .then(
      res => {
        dispatch({ type: GET_SUMMARY_DOCUMENT, payload: res.data });
      },
      error => {
        if (isError(error)) {
          dispatch(showFullPageError());
        } else {
          dispatch(unauthorised());
        }
      });

    function unauthorised() { return { type: GET_SUMMARY_UNAUTHORISED }; }
};