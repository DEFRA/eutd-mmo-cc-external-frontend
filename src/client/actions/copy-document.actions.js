import { showFullPageError, apiCallFailed, isError, CLEAR_ERRORS } from './index';
export const CONFIRM_COPY_DOCUMENT            = 'confirm_copy_document';
export const UNAUTHORISED_COPY_DOCUMENT       = 'unauthorised_copy_document';
export const CLEAR_COPY_DOCUMENT              = 'clear_copy_document';
export const REMOVE_CONFIRM_COPY_CERTIFICATE  = 'remove_confirm_copy_certificate';
export const ADD_ERROR_COPY_CERTIFICATE       = 'add_error_copy_certificate';
export const REMOVE_ERROR_COPY_CERTIFICATE    = 'remove_error_copy_certificate';

export const submitCopyCertificate = (payload) => async (dispatch, _getState, {orchestrationApi}) => {
  const {documentNumber, copyDocumentAcknowledged, voidOriginal, excludeLandings, journey} = payload;
  const config = {headers: {documentnumber: documentNumber}};

  try {
    let ipAddress;
    if (voidOriginal) {
      const ip = await orchestrationApi.get('/client-ip');
      ipAddress = ip.data;
    }

    return await orchestrationApi.post('/confirm-copy-certificate', { copyDocumentAcknowledged, voidOriginal, excludeLandings, journey, ipAddress }, config)
    .then(res => {
      dispatch({ type: CONFIRM_COPY_DOCUMENT, payload: {
        documentNumber: res.data.documentNumber,
        copyDocumentAcknowledged: res.data.copyDocumentAcknowledged,
        voidDocumentConfirm: res.data.voidOriginal
      }});
      return res.data.newDocumentNumber;
    });
  } catch (e) {
    if (isError(e)) dispatch(showFullPageError());
    if (e.response && e.response.status === 403) {
      dispatch({ type: UNAUTHORISED_COPY_DOCUMENT }); return;
    } else {
      dispatch(apiCallFailed(e.response));
    }
  }
};

export const checkCopyCertificate = (documentNumber) => async (dispatch, _getState, {orchestrationApi}) => {
  const config = {
    headers: {
      documentnumber: documentNumber
    }
  };

  try {
    const res = await orchestrationApi.get('/check-copy-certificate', config);
    const { canCopy } = res.data;
    if (!canCopy) {
      dispatch(unauthorised());
    }
  } catch(e) {
    if (isError(e)) dispatch(showFullPageError());
    if (e.response && e.response.status === 403) {
      dispatch(unauthorised());
      throw new Error(`Unauthorised access ${e.response.status}`);
    } else {
      dispatch(apiCallFailed(e.response));
      throw new Error(`An error has occurred ${e.response.status} (${e.response.statusText})`);
    }
  }

  function unauthorised() { return { type: UNAUTHORISED_COPY_DOCUMENT }; }
};

export const addCopyDocument = (data) => (dispatch) => dispatch({ type: CONFIRM_COPY_DOCUMENT, payload: data });
export const clearErrors = () => (dispatch) =>  dispatch({ type: CLEAR_ERRORS });
export const clearCopyDocument = () => (dispatch) => dispatch({ type: CLEAR_COPY_DOCUMENT });
export const unauthorisedCopyDocument = () => (dispatch) => dispatch({ type: UNAUTHORISED_COPY_DOCUMENT });