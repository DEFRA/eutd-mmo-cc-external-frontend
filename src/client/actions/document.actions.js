// actions
import {isError, showFullPageError} from './index';

export const GOT_ALL_DOCUMENTS          = 'GOT_ALL_DOCUMENTS';
export const GET_ALL_DOCUMENTS_FAILED   = 'GET_ALL_DOCUMENTS_FAILED';
export const GET_COMPLETED_DOCUMENT = 'GET_COMPLETED_DOCUMENT';
export const GET_PENDING_DOCUMENT = 'GET_PENDING_DOCUMENT';


// action creators
export const getAllDocuments = (type, month, year) => async (dispatch, getState, { orchestrationApi }) => {
  try {
    const res = await orchestrationApi.get(`/documents/${year}/${month}?type=${type}`);
    return dispatch({ type: GOT_ALL_DOCUMENTS, payload: res.data });

  } catch(e) {
    if( isError(e) ) dispatch(showFullPageError());
    //throw new Error(e.response);
  }
};

export const getCompletedDocument = (documentNumber) => async (dispatch, getState, { orchestrationApi }) => {
  try {
    const res = await orchestrationApi.get(`/document/${documentNumber}`);
    if (res.data && res.data.documentStatus === 'COMPLETE') {
      return dispatch({ type: GET_COMPLETED_DOCUMENT, payload: res.data });
    }
    return dispatch({ type: GET_COMPLETED_DOCUMENT, payload: null });
  }
  catch (e) {
    return dispatch({ type: GET_COMPLETED_DOCUMENT, payload: null });
  }
};

export const getPendingDocument = (documentNumber) => async (dispatch, getState, { orchestrationApi }) => {
  try {
    const res = await orchestrationApi.get(`/document/${documentNumber}`);
    if (res.data && res.data.documentStatus === 'PENDING') {
      return dispatch({ type: GET_PENDING_DOCUMENT, payload: res.data });
    }
    return dispatch({ type: GET_PENDING_DOCUMENT, payload: null });
  }
  catch (e) {
    return dispatch({ type: GET_PENDING_DOCUMENT, payload: null });
  }
};