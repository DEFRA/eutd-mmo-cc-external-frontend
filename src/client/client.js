// Startup point for the client side application
import 'babel-polyfill';
import React from 'react';
import ReactDOM from 'react-dom';
import {Router, withRouter} from 'react-router-dom';
import { createStore, applyMiddleware, compose } from 'redux';
import thunk from 'redux-thunk';
import { Provider } from 'react-redux';
import { renderRoutes } from 'react-router-config';
import axios from 'axios';
import Routes from './Routes';
import reducers from './reducers';
import { createBrowserHistory } from 'history';

const composeEnhancers = process.env.NODE_ENV === 'development' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__  ?
  window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ : compose;
  

const paramSerializerFn = params => {
  let output = '';
  Object.keys(params).forEach((k) => {
    output+= `${k}=${params[k]}&`;
  });
  return encodeURI(output);
};

const axiosResponseInterceptorSuccessFn = response => response;
const axiosResponseInterceptorErrorFn = error => {
  if (error.response.status === 599) {
    // Let's check if this is because we're logged out
    const empPromise = new Promise(() => {});
    const searchParam = error.request.responseURL.slice(error.request.responseURL.search(/[?]/g));
    let searchParams = new URLSearchParams(searchParam);
    if(searchParams.get('notLoggedInErr') === 'yes') {
      window.location = '/logout?backToPath=/server-logout';
      return empPromise;
    }
    // It's always a timeout...
    window.location = '/timed-out';
    return new Promise(() => {});
  } else {
    return new Promise((_,rej) => rej(error));
  }
};

const axiosInstanceForOrchService = axios.create({
  baseURL: '/orchestration/api/v1/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForOrchService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const axiosInstanceForRefService = axios.create({
  baseURL: '/reference/api/v1/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForRefService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const axiosInstanceForDynamixService = axios.create({
  baseURL: '/dynamix/',
  paramsSerializer: paramSerializerFn,
  timeout: window.INITIAL_STATE.config.axiosTimeout
});

axiosInstanceForDynamixService.interceptors.response.use(axiosResponseInterceptorSuccessFn, axiosResponseInterceptorErrorFn);

const store = createStore(
  reducers,
  window.INITIAL_STATE,
  composeEnhancers(applyMiddleware(
    // The difference between orch/ref service api is not significant as they both go through proxy
    // However it is significant in the server side store. To stick with same action creators we're
    // passing down same axios instance to both orchestration and reference service api .
    thunk.withExtraArgument( {
      orchestrationApi: axiosInstanceForOrchService,
      referenceServiceApi: axiosInstanceForRefService,
      dynamixApi: axiosInstanceForDynamixService } )
  ))
);

const ScrollToTop = withRouter( class ScrollToTop extends React.Component {
  componentDidMount() {
  }

  render() {
    return this.props.children;
  }
});

const history = createBrowserHistory();


ReactDOM.hydrate(
  <Provider store={store}>
    <Router history={history}>
      <ScrollToTop>
        <div>{renderRoutes(Routes)}</div>
      </ScrollToTop>
    </Router>
  </Provider>,
  document.querySelector('#root')
);
