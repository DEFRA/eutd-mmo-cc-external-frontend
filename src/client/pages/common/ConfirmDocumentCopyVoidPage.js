import React from 'react';
import {withRouter} from 'react-router';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import _ from 'lodash';
import SecondaryButton from '../../components/elements/SecondaryButton';
import PageTemplateWrapper from '../../components/PageTemplateWrapper';
import { withTranslation } from 'react-i18next';

import {
  dispatchApiCallFailed
} from '../../actions';

import {
  addCopyDocument,
  submitCopyCertificate,
  clearErrors,
} from '../../actions/copy-document.actions';

import {
  Main,
  Header,
  BackLink,
  GridRow,
  GridCol,
  MultiChoice
} from 'govuk-react';

import PageTitle from '../../components/PageTitle';
import Form from '../../components/elements/Form';
import HelpLink from '../../components/HelpLink';
import ContinueButton from '../../components/elements/ContinueButton';
import ErrorIsland from '../../components/elements/ErrorIsland';
import { scrollToErrorIsland, onHandleErrorClick } from '../utils';
import SelectRadio from '../../components/elements/SelectRadio';

class ConfimDocumentCopyVoidPage extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      voidOriginal: undefined,
      disableSubmit: undefined
    };
  }

  onChange = e => {
    this.setState({ voidOriginal: JSON.parse(e.target.value)});
  };

  onSubmit = async e => {
    e.preventDefault();
    this.setState({ disableSubmit: true });

    try {
      const { journey, nextUri} = this.props.route;
      const { documentNumber } = this.props.match.params;
      const { history, copyDocumentAcknowledged, copyExcludeLandings } = this.props;
      const { voidOriginal } = this.state;
      const confirmVoidOriginal = await this.props.submitCopyCertificate({
        journey,
        documentNumber,
        voidOriginal,
        copyDocumentAcknowledged,
        excludeLandings: copyExcludeLandings
      });

      if (confirmVoidOriginal) {
        history.push(nextUri.replace(':documentNumber', confirmVoidOriginal));
      }
    } catch (error) {
      scrollToErrorIsland();
    } finally {
      this.setState({ disableSubmit: false });
    }
  };

  componentDidMount = () => {
    window.scrollTo(0, 0);
    const { copyDocumentAcknowledged, voidDocumentConfirm } = this.props;

    if (!copyDocumentAcknowledged || !voidDocumentConfirm) {
       this.unauthorised();
    }
  }

  componentDidUpdate() {
    const { unauthorised } = this.props;

    if (unauthorised) {
      this.unauthorised();
    }
  }

  componentWillUnmount() {
    this.props.clearErrors();
  }

  unauthorised() {
    this.props.history.push('/forbidden');
  }

  render() {
    const {voidOriginalError, errors} = this.props.errors;
    const {route, match, unauthorised, history, t} = this.props;
    const { disableSubmit } = this.state;

    const {path, previousUri, journeyText, journey, cancelUri} = route;

    const title = t(`${journey}CopyVoidConfirmationPagetitle`, {journeyText: t(journeyText)})
    const h1Text =  t(`${journey}CopyVoidConfirmationHeader`, {journeyText: t(journeyText)})
    const currentUri = path.replace(':documentNumber', match.params.documentNumber);

    if (unauthorised)
       return null;

    return (
      <Main>
        {errors && (
          <ErrorIsland
            errors={errors.map(err => ({
              message: t(err.text),
              key: err.targetName
            }))}
            onHandleErrorClick={onHandleErrorClick}
        />
        )}

        <PageTitle
          title={`${!_.isEmpty(errors) ? 'Error: ' : ''}${title} - ${t(this.props.route.title)}`}
        />
        <GridRow>
          <GridCol columnTwoThirds>
            <BackLink onClick={e => {
              e.preventDefault();
              history.push(previousUri.replace(':documentNumber', match.params.documentNumber));
            }} href={previousUri}>
              {t('commonBackLinkBackButtonLabel')}
            </BackLink>
          </GridCol>
        </GridRow>

        <Form action='/orchestration/api/v1/confirm-copy-certificate' currentUrl={currentUri} onSubmit={this.onSubmit}>
          <GridRow>
            <GridCol style={{ width: '75%' }}>
              <GridRow>
                <GridCol style={{ width: '75%' }}>
                  <Header level='1' style={{ marginBottom: '0px' }}>
                    <legend>{h1Text}</legend>
                  </Header>
                </GridCol>
              </GridRow>
              <MultiChoice meta={{touched: t(voidOriginalError), error: t(voidOriginalError)}} id='voidOriginal'>
                <fieldset className="govuk-fieldset">
                  <SelectRadio id="documentVoidOriginalYes" checked={this.state.voidOriginal === true} value='true' name='voidOriginal' onChange={this.onChange}>{t(`${journey}CopyVoidConfirmationDocumentVoidOriginalYes`, {journeyText: t(journeyText).toLowerCase()})}</SelectRadio >
                  <SelectRadio id="documentVoidOriginalNo" checked={this.state.voidOriginal === false} value='false' name='voidOriginal' onChange={this.onChange}>{t(`${journey}CopyVoidConfirmationDocumentVoidOriginalNo`, {journeyText: t(journeyText).toLowerCase()})}</SelectRadio>
                </fieldset>
              </MultiChoice>
            </GridCol>
          </GridRow>
          <GridRow>
            <GridCol columnTwoThirds>
            <SecondaryButton
              id="cancel-confirm-copy-void-btn"
              formAction={null}
              name="cancel-confirm-copy-void-btn"
              onClick={(e) => {
                e.preventDefault();
                history.push(cancelUri);
              }}
            >
              {t('commonSecondaryButtonCancelButton')}<span className="govuk-visually-hidden">{t('commonSecondaryButtonCancelButton')}</span>
            </SecondaryButton>
              <ContinueButton type='submit' id='continue' disabled={disableSubmit}>{t('commonContinueButtonContinueButtonText')}</ContinueButton>
            </GridCol>
          </GridRow>
        </Form>
        <HelpLink journey={route.journey} />
      </Main>
    );
  }
}

function mapStateToProps(state) {
  return {
    errors: state.errors,
    unauthorised: state.confirmCopyDocument.unauthorised,
    copyDocumentAcknowledged: state.confirmCopyDocument.copyDocumentAcknowledged,
    copyExcludeLandings: state.confirmCopyDocument.copyExcludeLandings,
    voidDocumentConfirm: state.confirmCopyDocument.voidDocumentConfirm
  };
}

function loadData(store) {
  if (Object.keys(this.queryParams).length !== 0) {
    store.dispatch(dispatchApiCallFailed(JSON.parse(this.queryParams.error)));
  }
}

export const component = withRouter(
  connect(
    mapStateToProps,
    {
      addCopyDocument,
      submitCopyCertificate,
      clearErrors,
    }
  )(withTranslation() (ConfimDocumentCopyVoidPage))
);

ConfimDocumentCopyVoidPage.propTypes = {
  errors: PropTypes.object
};

export default {
  loadData,
  component: PageTemplateWrapper(component),
};
