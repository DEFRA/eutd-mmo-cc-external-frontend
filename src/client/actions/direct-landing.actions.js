import { beginApiCall, apiCallFailed, isForbidden, showFullPageError, CLEAR_ERRORS, isSilverLineError } from './index';
import { exportPayloadActionTypes } from '../actions';


export const GET_DIRECT_LANDING_PRODUCTS = 'get_direct_landing_products';
export const CLEAR_DIRECT_LANDING = 'clear_direct_landing';
export const DIRECT_LANDING_PRODUCTS_FAILED = 'direct_landing_products_failed';
export const POST_DIRECT_LANDING_PRODUCTS_UNAUTHORISED = 'post_direct_landing_products_unauthorised';
export const GET_DIRECT_LANDING_PRODUCTS_UNAUTHORISED = 'get_direct_landing_products_unauthorised';
export const POST_DIRECT_LANDING_PRODUCTS_WAF_ERROR = 'post_direct_landing_prodcts_waf_error';

export const getDirectLandings =
  (documentNumber) =>
  async (dispatch, _getState, { orchestrationApi }) => {
    const config = { headers: { documentnumber: documentNumber } };

    try {
      dispatch(beginApiCall());

      const res = await orchestrationApi.get(
        '/export-certificates/export-payload/direct-landings',
        config
      );

      await dispatch({ type: GET_DIRECT_LANDING_PRODUCTS, payload: res.data });
    } catch (e) {
      if (isForbidden(e)) {
        dispatch({ type: GET_DIRECT_LANDING_PRODUCTS_UNAUTHORISED });
      } else {
        dispatch(apiCallFailed(e.response));
        throw new Error(
          `An error has occurred ${e.response.status} (${e.response.statusText})`
        );
      }
    }
  };

export const clearDirectLanding = () => ({ type: CLEAR_DIRECT_LANDING });

export const upsertDirectLanding = (landing, documentNumber = undefined) => (dispatch, getState, { orchestrationApi } ) => {
  const config = (documentNumber)
    ? {
        headers: {
          documentnumber: documentNumber
        }
    }
    : {};

  return orchestrationApi.post('/export-certificates/direct-landing/validate', landing, config)
    .then(
      res => {
        dispatch(landingUpserted(res.data));
        dispatch(clearErrors());
      },
      error => {
        if (error.response.status === 400) {
          dispatch(landingUpserted(error.response.data));
          throw new Error();
        } else if (isSilverLineError(error)) {
          dispatch({ type: POST_DIRECT_LANDING_PRODUCTS_WAF_ERROR, supportID: error.response.data.match(/\d+/)[0] });
        } else if (isForbidden(error)) {
          dispatch({ type: POST_DIRECT_LANDING_PRODUCTS_UNAUTHORISED });
        } else {
          dispatch(showFullPageError());
        }
      });

  function landingUpserted(exportPayload) { return { type: exportPayloadActionTypes.LANDING_UPSERTED, exportPayload, errors: exportPayload.errors }; }
  function clearErrors() { return { type: CLEAR_ERRORS }; }
};
