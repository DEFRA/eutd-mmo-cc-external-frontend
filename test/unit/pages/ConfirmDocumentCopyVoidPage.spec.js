import _ from 'lodash';
import { mount } from 'enzyme/build';
import { Provider } from 'react-redux';
import * as React from 'react';
import { act } from 'react-dom/test-utils';
import { component as ConfirmDocumentCopyVoidPage } from '../../../src/client/pages/common/ConfirmDocumentCopyVoidPage';
import copyDocumentCopyVoidPage from '../../../src/client/pages/common/ConfirmDocumentCopyVoidPage';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { MemoryRouter } from 'react-router-dom';
import { addCopyDocument, clearCopyDocument, submitCopyCertificate, clearErrors } from '../../../src/client/actions/copy-document.actions';
import { dispatchApiCallFailed } from '../../../src/client/actions';
import { render } from '@testing-library/react';
jest.mock('../../../src/client/actions');
jest.mock('../../../src/client/actions/copy-document.actions');

const mockStore = configureStore([thunk.withExtraArgument({
  orchestrationApi: {
    post: () => {
      return new Promise((res) => {
        res({});
      });
    }
  }
})]);

const props = {
  errors: { errors: [], voidOriginalError: [] },
  match: {
    path: '/create-catch-certificate/:documentNumber/copy-void-confirmation',
    url: '/create-catch-certificate/GBR-2021-CC-D6FBF748C/copy-void-confirmation',
    isExact: true,
    params: {'documentNumber': 'GBR-2021-CC-D6FBF748C'}
  },
  route: {
    title                 : 'Create a UK catch certificate - GOV.UK',
    journey               : 'catchCertificate',
    previousUri           : '/create-catch-certificate/GBR-2021-CC-D6FBF748C/copy-this-catch-certificate',
    path                  : '/create-catch-certificate/:documentNumber/copy-void-confirmation',
    journeyText           : 'catch certificate',
    nextUri: '/create-catch-certificate/:documentNumber/add-your-reference'
  },
  t: jest.fn()
};

const storeA = {
  confirmCopyDocument: {
    copyDocumentAcknowledged: true,
    copyExcludeLandings: false,
    unauthorised: false,
    voidOriginal: undefined,
    voidDocumentConfirm: false
  },
  errors: { errors: [] },
  t: jest.fn()
};

let wrapper;

describe('Confirm Document Copy Void page', () => {

  beforeEach(() => {
    addCopyDocument.mockReturnValue({ type: 'CONFIRM_COPY_DOCUMENT' });
    clearCopyDocument.mockReturnValue({ type: 'CLEAR_COPY_DOCUMENT' });
    submitCopyCertificate.mockResolvedValue('new-document-number');
    dispatchApiCallFailed.mockResolvedValue('CALL_FAILED');
    clearErrors.mockReturnValue({ type: 'CLEAR_ERRORS '});

    window.scrollTo = jest.fn();

    wrapper = mount(
      <Provider store={mockStore(storeA)}>
        <MemoryRouter>
          <ConfirmDocumentCopyVoidPage {...props} />
        </MemoryRouter>
      </Provider>
    );
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should load successfully', () => {
    expect(wrapper).toBeDefined();
  });

  it('should render a continue button', () => {
    let continueButton = wrapper.find({type: 'submit', id: 'continue'});
    expect(continueButton.exists()).toBeTruthy();
  });

  it('should render a cancel button', () => {
    let continueButton = wrapper.find({name: 'cancel-confirm-copy-void-btn', id: 'cancel-confirm-copy-void-btn'});
    expect(continueButton.exists()).toBeTruthy();
  });

  it('should call the onClick of the back link', () => {
    const mockPreventDefault = jest.fn();

    const baclLinkEl = wrapper.find('BackLink');
    baclLinkEl.simulate('click', {
      preventDefault: mockPreventDefault
    });

    expect(mockPreventDefault).toHaveBeenCalled();
  });


  it('should call the onClick of the Secondary Button', () => {
    const mockPreventDefault = jest.fn();

    const secondaryButtonEl = wrapper.find('button[name="cancel-confirm-copy-void-btn"]');
    secondaryButtonEl.simulate('click', {
      preventDefault: mockPreventDefault
    });

    expect(mockPreventDefault).toHaveBeenCalled();
  });

  it('should call submitCopyCertificate when the form is submitted', () => {
    wrapper.find('form').simulate(
      'submit',
      {preventDefault() {}}
    );

    expect(submitCopyCertificate).toHaveBeenCalledTimes(1);
  });

  it('should default the submit button as not being disabled', () => {
    const button = wrapper.find({type: 'submit', id: 'continue'}).at(0);

    expect(button.prop('disabled')).toBe(false);
  });

  it('should set the submit button as disabled on submit', () => {
    const resolveAfterOneSec = () =>
      new Promise((res, _) => setTimeout(() => res(), 1000));

    submitCopyCertificate.mockImplementation(() => resolveAfterOneSec);

    wrapper.find('form').simulate(
      'submit',
      {preventDefault() {}}
    );

    const button = wrapper.find({type: 'submit', id: 'continue'}).at(0);

    expect(button.prop('disabled')).toBe(true);
  });

  it('should set the submit button back to enabled after successful submission', () => {
    wrapper.find('form').simulate(
      'submit',
      {preventDefault() {}}
    );

    const button = wrapper.find({type: 'submit', id: 'continue'}).at(0);

    expect(button.prop('disabled')).toBe(false);
  });

  it('should set the submit button back to enabled after failed submission', () => {
    submitCopyCertificate.mockRejectedValue('no');

    wrapper.find('form').simulate(
      'submit',
      {preventDefault() {}}
    );

    const button = wrapper.find({type: 'submit', id: 'continue'}).at(0);

    expect(button.prop('disabled')).toBe(false);
  });

  it('should handle submit event and push history', () => {
    act(() => wrapper.find('form').props().onSubmit(
      {preventDefault() {}}
    ));
  });

  it('should handle on change events', () => {
    wrapper.find('input[name="voidOriginal"][id="documentVoidOriginalYes"]').simulate('change', {target: {name: 'voidOriginal',  value: 'true'}});
  });

  it('should find the legend tags', () => {
    expect(wrapper.find('legend')).toHaveLength(1);
  });

  it('should throw an error if the user has not made a selection', () => {
    const store = _.cloneDeep(storeA);

    store.errors = {
      voidOriginal: undefined,
      errors: [{ 'targetName': 'voidOriginal', 'text': 'Select an option to continue' }]
    };

    wrapper = mount(
      <Provider store={mockStore(store)}>
        <MemoryRouter>
          <ConfirmDocumentCopyVoidPage {...props} />
        </MemoryRouter>
      </Provider>
    );

    let ErrorIsland = wrapper.find('div#errorIsland');
    let errorSummary = wrapper.find('ul.error-summary-list li a');
    expect(ErrorIsland.exists()).toBeTruthy();
    expect(errorSummary.text()).toEqual('Select an option to continue');
  });

  it('should push history for component did update', () => {
    const mockPush = jest.fn();

    new ConfirmDocumentCopyVoidPage.WrappedComponent({
      match: {
        params: {
          documentNumber: '',
        },
      },
      unauthorised: true,
      history: {
        push: mockPush,
      },
    }).componentDidUpdate();

    expect(mockPush).toHaveBeenCalled();
  });

  it('should clear errors on component will unmount', () => {
    const mockDispatchClearErrors = jest.fn();

    new ConfirmDocumentCopyVoidPage.WrappedComponent({
      match: {
        params: {
          documentNumber: ''
        }
      },
      unauthorised: true,
      clearErrors: mockDispatchClearErrors,
    }).componentWillUnmount();

    expect(mockDispatchClearErrors).toHaveBeenCalled();
  });
});

describe('loadData', () => {
  const store = {
    dispatch: jest.fn()
  };

  const documentNumber = 'some-document-number';
  const queryParams = {
    error: '{"x":5,"y":6}'
  };

  it('will call all methods needed to load the component', () => {
    copyDocumentCopyVoidPage.documentNumber = documentNumber;
    copyDocumentCopyVoidPage.queryParams = {};

    copyDocumentCopyVoidPage.loadData(store);

    expect(store.dispatch).not.toHaveBeenCalled();
  });

  it('will call all methods needed to load the component when given query params', () => {
    copyDocumentCopyVoidPage.documentNumber = documentNumber;
    copyDocumentCopyVoidPage.queryParams = queryParams;

    copyDocumentCopyVoidPage.loadData(store);

    expect(store.dispatch).toHaveBeenCalled();
  });
});

describe('snapshots describe', () => {
  beforeEach(() => {
   clearErrors.mockReturnValue({ type: 'CLEAR_ERRORS '});
    window.scrollTo = jest.fn();
  });

  it('should load successfully', () => {
    expect(wrapper).toBeDefined();
  });

  it('should take a snapshot of the whole page', () => {
    const { container } = render(wrapper);
    expect(container).toMatchSnapshot();
  })
})

