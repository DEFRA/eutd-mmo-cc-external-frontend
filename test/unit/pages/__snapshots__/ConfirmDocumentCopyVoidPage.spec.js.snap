// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshots describe should take a snapshot of the whole page 1`] = `
<div>
  <div
    class="css-1l99ixu-OuterContainer e4u1u2g0"
  >
    <div
      class="css-n57fe4-InnerContainer e4u1u2g1"
    >
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-n58wva-StyledColumn e4u1u2g0"
        >
          <a
            class="e1vcssft0 css-zk98nb-Anchor-StyledHoc e4u1u2g0"
            href="/create-catch-certificate/GBR-2021-CC-D6FBF748C/copy-this-catch-certificate"
          >
            Back
          </a>
        </div>
      </div>
      <form
        action="/orchestration/api/v1/confirm-copy-certificate?c=/create-catch-certificate/undefined/copy-void-confirmation"
        method="POST"
      >
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <div
            class="css-1w80l5h-StyledColumn e4u1u2g0"
            style="width: 75%;"
          >
            <div
              class="css-1lzr9ic-StyledRow e4u1u2g0"
            >
              <div
                class="css-1w80l5h-StyledColumn e4u1u2g0"
                style="width: 75%;"
              >
                <h1
                  class="e1vcssft0 css-11nzzut-StyledHeader-StyledHoc e4u1u2g0"
                  style="margin-bottom: 0px;"
                >
                  <legend>
                    Are you sure you want to void the original catch certificate?
                  </legend>
                </h1>
              </div>
            </div>
            <div
              class="e1vcssft0 css-1phgi8s-StyledFieldset-StyledHoc e4u1u2g0"
              id="voidOriginal"
            >
              <span
                class="e1vcssft0 css-3e01s4-StyledLabelText-StyledHoc e4u1u2g0"
              />
              
              <fieldset
                class="govuk-fieldset"
              >
                <label
                  class="css-sdrcg6-StyledHoc"
                  for="documentVoidOriginalYes"
                  id="label-documentVoidOriginalYes"
                >
                  <input
                    class="css-1xt3i3l"
                    id="documentVoidOriginalYes"
                    name="voidOriginal"
                    type="radio"
                    value="true"
                  />
                  <span
                    class="css-1uwxoxa"
                  >
                    Yes, create a copy of the catch certificate then void the original
                  </span>
                </label>
                <label
                  class="css-sdrcg6-StyledHoc"
                  for="documentVoidOriginalNo"
                  id="label-documentVoidOriginalNo"
                >
                  <input
                    class="css-1xt3i3l"
                    id="documentVoidOriginalNo"
                    name="voidOriginal"
                    type="radio"
                    value="false"
                  />
                  <span
                    class="css-1uwxoxa"
                  >
                    No, create a copy of the catch certificate only
                  </span>
                </label>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="css-1lzr9ic-StyledRow e4u1u2g0"
        >
          <div
            class="css-n58wva-StyledColumn e4u1u2g0"
          >
            <button
              class="css-17uw1q2-StyledButton-StyledHoc e4u1u2g0"
              id="cancel-confirm-copy-void-btn"
              name="cancel-confirm-copy-void-btn"
            >
              Cancel
              <span
                class="govuk-visually-hidden"
              >
                Cancel
              </span>
            </button>
            <button
              class="css-1gc9650-StyledButton-StyledHoc e4u1u2g0"
              id="continue"
              type="submit"
            >
              Continue
            </button>
          </div>
        </div>
      </form>
      <div
        class="css-1lzr9ic-StyledRow e4u1u2g0"
      >
        <div
          class="css-1w80l5h-StyledColumn e4u1u2g0"
        >
          <hr
            class="help-link"
          />
          <h2
            class="e1vcssft0 css-162lx7h-StyledHeader-StyledHoc e4u1u2g0"
          >
            Need help?
          </h2>
          <a
            href="undefined"
            rel="noopener noreferrer"
            target="_blank"
          >
            Get help exporting fish from the UK (gov.uk)
             
            <span
              class="govuk-visually-hidden"
            >
              (opens in new tab)
            </span>
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
